### PR TITLE
Create runtime implementation

### PR DESCRIPTION
**Pull request by rbren**
First pass at moving all the action-running logic into a Runtime class.

The Runtime handles all the logic that interacts with the external world: browsing, reading/writing files, running commands, etc.

There are still some actions (e.g. Finish, Delegate, ModifyTask) that are handled internally by the controller. That's all in one place now.

The Action classes themselves are now pure data classes, with no internal logic.

TODO:
* [x] ~~Reimplement GitHub action~~ (spoke with Graham, we'll reimplement this a different way)
* [x] Create E2B runtime

### CODE REVIEW

File: opendevin/events/action/commands.py [hunk 1]
```diff
@@ -60,25 +50,8 @@ class IPythonRunCellAction(Action):
...
-            config.get(ConfigType.WORKSPACE_MOUNT_PATH_IN_SANDBOX),
-            '.tmp',
-            '.ipython_execution_tmp.py',
-        )
-        obs = controller.action_manager.run_command(
-            f'execute_cli < {tmp_filepath_inside_sandbox}', background=False
-        )
-        return IPythonRunCellObservation(content=obs.content, code=self.code)
+    def runnable(self) -> bool:
+        return True
```
Comment (Umpire2018) [inline opendevin/events/action/commands.py:54]: Perhaps we could also simplify the code by using class attributes directly instead of properties.

Comment (rbren) [reply opendevin/events/action/commands.py:54]: I would love that, but managing class attributes with `dataclass` and subclasses seems to be impossible. There's no way for me to add a `runnable = False` to the base class without getting a bunch of errors about default args being set before non-defaults

File: opendevin/runtime/runtime.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            return ErrorObservation(f'Action {action_id} does not exist.')
+        if not hasattr(self, action_id):
+            return ErrorObservation(
+                f'Action {action_id} is not supported in the current runtime.'
+            )
+        print('actually running')
+        observation = await getattr(self, action_id)(action)
+        return observation
+
+    def get_background_obs(self) -> List[CmdOutputObservation]:
```
Comment (neubig) [inline opendevin/runtime/runtime.py:79]: A docstring would be good here.

**Review by neubig — APPROVED** [review]
Basically looks great! I left a few suggestions for improvement, but I approve merging.

File: opendevin/runtime/runtime.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    DockerSSHBox,
+    E2BBox,
+    LocalBox,
+    Sandbox,
+)
+from opendevin.runtime.browser.browser_env import BrowserEnv
+from opendevin.runtime.plugins import PluginRequirement
+
+
+def get_sandbox(sid: str = 'default', sandbox_type: str = 'exec') -> Sandbox:
```
[the remark this answers is not in the document]
Comment (rbren) [reply opendevin/runtime/runtime.py:34]: good call!

File: opendevin/runtime/runtime.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        self.sid = sid
+        sandbox_type = config.get(ConfigType.SANDBOX_TYPE)
+        self.sandbox = get_sandbox(sid, sandbox_type)
+        self.browser = BrowserEnv()
+
+    def init_sandbox_plugins(self, plugins: List[PluginRequirement]):
+        self.sandbox.init_plugins(plugins)
+
+    async def run_action(self, action: Action) -> Observation:
+        print('run action!', action.runnable)
```
[the remark this answers is not in the document]
Comment (rbren) [reply opendevin/runtime/runtime.py]: whoops leftover debugging 😄

File: opendevin/events/action/commands.py (hunk 1 above)
Comment (enyst) [reply opendevin/events/action/commands.py:54]: Addressed by https://github.com/OpenDevin/OpenDevin/pull/1679

File: opendevin/runtime/server/runtime.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        working_dir = self.sandbox.get_working_directory()
+        return await write_file(
+            action.path, working_dir, action.content, action.start, action.end
+        )
+
+    async def browse(self, action: BrowseURLAction) -> Observation:
+        return await browse(action, self.browser)
+
+    async def recall(self, action: AgentRecallAction) -> Observation:
+        return NullObservation('')
```
Comment (SmartManoj) [inline opendevin/runtime/server/runtime.py:74]: Need to add a TODO comment?

**Pull request merged**